typescript/packages/jumble/src/views/CharmList.tsx

```diff
@@ -2,17 +2,63 @@ import { useCell } from "@/hooks/use-cell";
 import { NavLink } from "react-router-dom";
 import { NAME, UI } from "@commontools/builder";
 import { useCharmManager } from "@/contexts/CharmManagerContext";
-import { castNewRecipe } from "@commontools/charm";
+import { castNewRecipe, Charm } from "@commontools/charm";
 import { charmId } from "@/utils/charms";
-import { useEffect, useRef } from "react";
+import { useCallback, useEffect, useRef, useState } from "react";
 import { Card } from "@/components/Card";
 import { useParams } from "react-router-dom";
+import { render } from "@commontools/html";
 
 export interface CommonDataEvent extends CustomEvent {
   detail: {
     data: any[];
   };
 }
+function CharmPreview({ charm, replicaName }: { charm: Charm; replicaName: string }) {
+  const previewRef = useRef<HTMLDivElement | null>(null);
+  const [isIntersecting, setIsIntersecting] = useState(false);
+
+  useEffect(() => {
+    if (!previewRef.current) return;
+
+    const observer = new IntersectionObserver(
+      ([entry]) => {
+        setIsIntersecting(entry.isIntersecting);
+      },
+      { threshold: 0 },
+    );
+
+    observer.observe(previewRef.current);
+    return () => observer.disconnect();
+  }, []);
+
+  useEffect(() => {
+    if (!previewRef.current || !isIntersecting) return;
+    const preview = previewRef.current;
+    const charmData = charm.cell.get()?.[UI];
+    if (!charmData) return;
+    preview.innerHTML = ""; // Clear any existing rendered content
+    const cancel = render(preview, charmData);
+    return cancel;
+  }, [charm, isIntersecting]);
```

Comment:
Rather than running the charm, we just yoink the $UI property and render it... which happens to create an iframe and a whole react app but it's isolated from the actual running charm.